src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/Factory/CustomAuthenticatorFactory.php

```diff
@@ -0,0 +1,56 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Bundle\SecurityBundle\DependencyInjection\Security\Factory;
+
+use Symfony\Component\Config\Definition\Builder\ArrayNodeDefinition;
+use Symfony\Component\Config\Definition\Builder\NodeDefinition;
+use Symfony\Component\DependencyInjection\ContainerBuilder;
+
+class CustomAuthenticatorFactory implements AuthenticatorFactoryInterface, SecurityFactoryInterface
+{
+    public function create(ContainerBuilder $container, string $id, array $config, string $userProvider, ?string $defaultEntryPoint)
+    {
+        throw new \LogicException('Custom authenticators are not supported in the old Security system.');
+    }
+
+    public function getPosition(): string
+    {
+        return 'pre_auth';
+    }
+
+    public function getKey(): string
+    {
+        return 'authenticator';
+    }
+
+    /**
+     * @param ArrayNodeDefinition $builder
+     */
+    public function addConfiguration(NodeDefinition $builder)
+    {
+        $builder
+            ->fixXmlConfig('service')
```

Comment:
Ultimately, custom config looks like this:

```yml
            authenticator:
                services:
                    - App\Security\TestingAuthenticator
```

I'm wondering if we should call the key `custom_authenticators`?

And... is there any need for the `services` sub-key? It could just be:

```yml
            custom_authenticator:
                - App\Security\TestingAuthenticator
```

Reply:
> And... is there any need for the services sub-key? It could just be:

From a PHP code of view, yes. The `custom_authenticator` node is created by `MainConfiguration`, so the factory only can control wants inside it (in theory). That's not a problem, but with `prototype` nodes it (as `fixXmlConfig()` needs to be called on the parent node).

I think we can - with lots of hackery - get the parent node of `$builder` and then make this node a prototype, but as I said that is very probably very error prone/hacky

Reply:
Sounds good - I'll leave this unresolved until we make the change to `custom_authenticator`.

Reply:
I'm :+1: for `custom_authenticator`. Does "sounds good" mean "let's try to hacky stuff" or "keep services for now"? :)

Reply:
Keep services - fine for me - something we could easily change later if we're motivated.